### ISSUE TITLE
Restrict OCM to ScienceMesh or not?

### ISSUE THREAD

**Issue by michielbdejong** [sciencemesh/sciencemesh#196]
Vision 1 (closed network):
ScienceMesh makes sure you never receive a share from a person who does not have an account at one of the participating sites, nor from a person whose invite you rejected or never received. In accordance, you will not be able to share with a user who does not have an account at a ScienceMesh site, or who has not accepted your invite.

Vision 2 (discovery only):
ScienceMesh makes it easy to establish a collaboration with someone at another ScienceMesh site; you just give them an invite through any channel you know, they open the URL with any browser and see a list in which they'll choose their home site. The receiver of the invite doesn't even need to bookmark or remember the URL of their EFSS, this is discoverable through meshdir from the invite link. However, if you know someone's full `user@domain.com` OCM address then you can also bypass the invite flow and use the share-with flow of OCM. If the other user is not at a ScienceMesh domain then it may be necessary to explicitly "Accept" an incoming share; this step can be skipped for ScienceMesh sites, because they can be marked as trusted servers.

**Comment by michielbdejong** [comment]
@glpatcern last week we discussed vision 1 but this morning when chatting with other people I thought maybe vision 2 is more appropriate. What do you think?

**Comment by glpatcern** [comment]
With more thinking together with @gmgigi96, there's still an issue with a totally open vision, and that is more of a "political" decision: do we want to trust unknown sites or not? Who checks that? "Quis custodiet ipsos custodes?"

Think about the following scenario:
I am a malicious actor. I implement a `https://badguycloud.com` domain with a Let's Encrypt certificate, and I even expose `https://badguycloud.com/ocm-provider` with all details and a public key. Then I send a signed share request (the signing is not yet implemented but let's pretend it is). The recipient can validate the request against the `/ocm-provider` metadata, but the system cannot "validate" a non trusted URL such as that without an explicit, manually managed whiltelist as provided by the ScienceMesh directory.

Therefore there is a value to at least allowing sites to block shares from unknown / untrusted domains.

**Comment by michielbdejong** [comment]
> allowing sites to block shares from unknown / untrusted domains

Yes, and this is already provided by both OC-10 and Nextcloud. There is a list of trusted domains, and shares coming in from untrusted servers are either rejected or queued for manual approval

**Comment by glpatcern** [comment]
Very good, and who maintains today the list of trusted domains? I guess each site is on your own in maintaining the list of trusted peers?

In such a case, ScienceMesh and its directory of sites would be an extra source of trusted domains. Now I understand what you say with:

> If the other user is not at a ScienceMesh domain then it may be necessary to explicitly "Accept" an incoming share; this step can be skipped for ScienceMesh sites, because they can be marked as trusted servers.

To which I'd add that a site may have a policy to _enforce_ users to explicitly "Accept" shares from untrusted sources. Or a site may automatically _reject_ such shares and manually configure a list of trusted servers. That is, I'd leave the standard open to both visions, and the decision is a site's administrator policy decision.

**Comment by michielbdejong** [comment]
I think we now have a consensus that it will always be a layered approach.

Invites really improve the security compared to the Wild Open Internet though.

### with invites
In the example above, for `user@badguycloud.com` to send me an invite, they would first need to have established an out-of-band communication with their victim. So they would need to send a phishing email or a fake contact request on a messaging app first. Two servers will only interact if there is a pre-established trust between at least one user on the one server and at least one user on the other server.

### without invites
In a flow without invites, they could directly send an me OCM share with a nice display name, and they would already be showing up in my EFSS GUI.
